[Fixes #8066] Added native ENS registration

jeluard opened the pull request:
fixes #8066

### Summary

Implement ENS registration according to latest Figma designs.

_Note to testers_  Local names persistency is not part of this PR. Thus initial welcome screen is always the same even if you already registered some names using native registration.

Address and public key will be connected to names, and stateofus names can be be registered.


status: WIP

dshulyak committed:
Add goerli testnet

Signed-off-by: Igor Mandrigin <i@mandrigin.ru>

hlolli committed:
[nix] patch realm as a nodePackage for linux and darwin and use it from nix store, add to default nix-shell buildDependencies, regenerate nodePackages with node2nix 1.6.1

rasom committed:
Send logs on shaking device

rasom committed:
App version and more detail in logs email

Commit:
[feature] use subscriptions for tokens

- removes fetching of last 100000 blocks of token transfers from
the wallet pull loop
- fetches the last 100000 blocks of token transfers at startup
- replaces pulling by subscriptions to ethlogs for token transfers

Commit:
[feature] use new block signal to get new transactions

- remove the transaction fetching loop entirely to rely only on subscription
for live transactions and token transfer updates
- fetch token transfers history via etherscan API to lift the 100000 blocks
limit on token transfers history

- inbound token transfers are catched via a filter on ethlogs
- outbound token transfers and other transactions are catched by filtering
transaction in current block that have the wallet address as to or from field

Commit:
[refactor] transaction details

Commit:
[refactor] transaction details

Commit:
[refactor] transaction history and filters

Commit:
[performance] improve wallet update performances

`wallet-autoconfig-token` is a very expensive call on mainnet
because it checks the balance of every known token.

it is called:
- when wallet is refreshed by pulling
- when user goes on any wallet screen

this PR changes that by:
- calling it only when the wallet is initialized and there is no
visible-token configuration

it only calls update-wallet when a new transaction arrives

Commit:
[performance] improve wallet update performances

`wallet-autoconfig-token` is a very expensive call on mainnet
because it checks the balance of every known token.

it is called:
- when wallet is refreshed by pulling
- when user goes on any wallet screen

this PR changes that by:
- calling it only when the wallet is initialized and there is no
visible-token configuration

it only calls update-wallet when a new transaction arrives

Commit:
[refactoring] remove web3, clean up wallet effects

- introduce json-rpc namespace, which provides `call` and `eth-call`,
a generic way of calling a json-rpc method taking care of conversions
and error handling
- remove web3 usage from wallet
- clean up effects, reducing the amount of computations when login in

Commit:
[refactor] remove ethereum `call` and `call-params`

- use `json-rpc/eth-call` and `json-rpc/eth-transaction-call`
everywhere
- move all conversions to abi-spec

Commit:
[refactor] remove ethereum `call` and `call-params`

- use `json-rpc/eth-call` and `json-rpc/eth-transaction-call`
everywhere
- move all conversions to abi-spec

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | fdfc033b | [#1](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-8248/1/) | 2019-05-21 12:51:44 | ~15 min | `android` | [:page_facing_up: log](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-8248/1/consoleText) |

status-im-auto commented:
### Jenkins Builds
| :grey_question: | Commit | :hash: | Finished (UTC) | Duration | Platform | Result |
|-|-|-|-|-|-|-|
| :x: | fdfc033b | [#1](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-8248/1/) | 2019-05-21 12:51:44 | ~15 min | `android` | [:page_facing_up: log](https://ci.status.im/job/status-react/job/prs/job/android/job/PR-8248/1/consoleText) |
| :x: | fdfc033b | [#1](https://ci.status.im/job/status-react/job/prs/job/linux/job/PR-8248/1/) | 2019-05-21 12:51:44 | ~15 min | `linux` | [:page_facing_up: log](https://ci.status.im/job/status-react/job/prs/job/linux/job/PR-8248/1/consoleText) |

jeluard committed:
[Fixes #8066] Added native ENS registration

jeluard committed:
[Fixes #8066] Added native ENS registration

jeluard committed:
[Fixes #8066] Added native ENS registration

errorists commented:
👏 Great 👏 work 👏 @jeluard 

All of the issues I could find, if you run into trouble with using fontWeight, try asking @yenda, he's the typography master 👑 

![Group 3](https://user-images.githubusercontent.com/4932774/58103372-24ff1400-7be3-11e9-9846-b9397224b14f.png)

---

![Group](https://user-images.githubusercontent.com/4932774/58103442-42cc7900-7be3-11e9-8b32-0201d3ecbbfd.png)

---

![Group 2](https://user-images.githubusercontent.com/4932774/58103447-46600000-7be3-11e9-9010-064c36fb047f.png)

---

![Artboard](https://user-images.githubusercontent.com/4932774/58103456-4a8c1d80-7be3-11e9-95d7-543fbeb8acca.png)

---

![Group 5](https://user-images.githubusercontent.com/4932774/58103458-4e1fa480-7be3-11e9-835a-d02bd4b24ddd.png)

---

![Group 4](https://user-images.githubusercontent.com/4932774/58103465-511a9500-7be3-11e9-94bd-b1684fdd191a.png)

yenda commented on src/status_im/utils/ethereum/stateofus.cljs:
```diff
@@ -0,0 +1,20 @@
+(defn- subdomain [username]
+  (str username "." domain))
+
+(def registrars
```
do we actually need this?

```
(ens/owner "0x112234455c3a32fd11230c42e7bccd4a84e02010" "stateofus.eth" println)
0x11d9f481effd20d76cee832559bd9aca25405841
```

jeluard replied:
Good point, will drop it

jeluard replied:
Actually it's something we need to check statically and pretty often so doesn't appear to be worth the hassle and perf cost.

yenda replied:
what do you mean we need to check it statically? what if we change the contract address for stateofus.eth?

jeluard replied:
Well same as if we change any contract address?

yenda replied:
@decanus do you have any recommendation here? Should we rather rely on hardcoded contract addresses for the registrar or solve it with the ens resolver contract?

yenda commented on src/status_im/utils/ethereum/ens.cljs:
```diff
@@ -129,4 +129,11 @@
             ens-name
             #(addr % ens-name cb)))
 
+(defn get-pubkey
```
this already exists in resolver.cljs

yenda replied:
It still already exists?

Commit:
[fix] always show tokens that have a balance

- temporary fix before visible assets are fixed
- the problem is that you can't hide an asset with the way it currently
works, it will come back every time as long as there is a balance for it

Signed-off-by: yenda <eric@status.im>

rasom committed:
RAM bundle basic setup

Commit:
[refactor] use `ethereum` `current-address` and `chain-keyword`

in many places in the codebase, we are doing various destructuring
and function calls to get the normalized current address and the chain
keyword for the current network

this PR replace all usages by utility functions introduced recently

Signed-off-by: yenda <eric@status.im>

flexsurfer commented:
@jeluard , @yenda introduced typography some time ago, so we don't need to send font styles manually anymore,  typography key should be used instead, you can see it name in figma cc @errorists

flexsurfer committed:
[#8239] App crash when submit cryptokitties with different than it's valid domain

Signed-off-by: Andrey Shovkoplyas <motor4ik@gmail.com>

flexsurfer commented:
![image](https://user-images.githubusercontent.com/11790366/58185161-c7d29380-7cb2-11e9-8d76-6b0f0cfa1657.png)

https://github.com/status-im/status-react/blob/496f3f1cc78703c3d629f9f458a0e966b04e5c61/src/status_im/ui/components/typography.cljs#L45

Commit:
[refactor] move utils.ethereum to ethereum

move utils.ethereum.tokens to ethereum.tokens
move utils.ethereum.abi-spec to ethereum.abi-spec
move utils.ethereum.core to ethereum.core
move utils.ethereum.eip165 to ethereum.eip165
move utils.ethereum.eip55 to ethereum.eip55
move utils.ethereum.eip681 to ethereum.eip681
move utils.ethereum.ens to ethereum.ens
move utils.ethereum.erc721 to ethereum.erc721
move utils.ethereum.mnemonics to ethereum.mnemonics
move utils.ethereum.resolver to ethereum.resolver
move utils.ethereum.macros to ethereum.macros

Signed-off-by: yenda <eric@status.im>

Commit:
[refactor] move utils.ethereum to ethereum

move utils.ethereum.tokens to ethereum.tokens
move utils.ethereum.abi-spec to ethereum.abi-spec
move utils.ethereum.core to ethereum.core
move utils.ethereum.eip165 to ethereum.eip165
move utils.ethereum.eip55 to ethereum.eip55
move utils.ethereum.eip681 to ethereum.eip681
move utils.ethereum.ens to ethereum.ens
move utils.ethereum.erc721 to ethereum.erc721
move utils.ethereum.mnemonics to ethereum.mnemonics
move utils.ethereum.resolver to ethereum.resolver
move utils.ethereum.macros to ethereum.macros

Signed-off-by: yenda <eric@status.im>

jeluard commented:
@flexsurfer @yenda I made some changes

flexsurfer commented on src/status_im/subs.cljs:
```diff
@@ -1633,3 +1634,27 @@
+
+(re-frame/reg-sub
+ :ens/state
+ (fn [db]
```
for root keys we need to use `reg-root-key-sub`

yenda replied:
and clearly here  there should be only one root key for ens and the rest is derived subscription. but looking at how they are used it should even rather be screen subscriptions that aggregate from an :ens root key and a few others

jeluard replied:
Not clear what you mean. Do you have an exemple?

yenda replied:
https://github.com/status-im/status-react/blob/develop/src/status_im/subs.cljs#L1199-1212 is a good exemple

Basically you need to think about each subscription you make as a cache, in your case you want to cache the :ens key of app-db as a root key, so that the subscriptions you use it in are only recomputed when the something in :ens is changed.

flexsurfer commented on src/status_im/subs.cljs:
```diff
@@ -1633,3 +1634,27 @@
+(re-frame/reg-sub
+ :ens/username
+ (fn [db]
+   (get-in db [:ens :username])))
```
`:ens` looks like too general, maybe we could use something like `:ens-usernames`

jeluard replied:
I think :ens is fine

yenda replied:
I agree :ens is fine but doing a get-in from the root app-db is not fine we are trying to optimize subscriptions now, this is going to get triggered for every change in app-db

flexsurfer commented on src/status_im/ui/screens/ens/events.cljs:
```diff
@@ -0,0 +1,123 @@
+(ns status-im.ui.screens.ens.events
```
i believe we don't use events in ui namespace, handlers should be in root events file, fx and functions in core.cljs

jeluard replied:
There are plenty events files.

yenda replied:
Which is wrong there shouldnt be any

Commit:
nix: Move debug realm into desktop folder and use RN cli from node_modulesSigned-off-by: Pedro Pombeiro <pombeirp@users.noreply.github.com>

dmitryn committed:
[#8256] do not call getApplicationInfo every time card connected

Signed-off-by: Dmitry Novotochinov <dmitry.novot@gmail.com>

yenda commented on src/status_im/ui/screens/ens/views.cljs:
```diff
@@ -0,0 +1,367 @@
+         [react/nested-text
+          {:style {:color colors/gray}}
+          (i18n/label :t/ens-welcome-point-1)
+          [{:style {:color colors/black :text-decoration-line :underline}}
```
black doesn't need to be specified it's the default

yenda commented on src/status_im/ui/screens/ens/events.cljs:
```diff
@@ -0,0 +1,123 @@
+ (fn [{db :db} [_ data]]
+   (let [transaction (prepare-extension-transaction data)]
+     ;; Last argument is nil: no need to estimate gas
+     (wallet/open-modal-wallet-for-transaction db transaction nil))))
```
use `wallet/eth-transaction-call` here

jeluard replied:
Can't find this function.

yenda replied:
It's in my PR which should get merged today

yenda commented on src/status_im/utils/ethereum/core.cljs:
```diff
@@ -108,6 +110,12 @@
   (when (= 66 (count s))
     (normalized-address (subs s 26))))
 
+(defn coordinates [public-key]
```
it is better to always put the param list of the second line, a good way to convince yourself is to write a docstring for this function

yenda commented on src/status_im/ui/screens/ens/views.cljs:
```diff
@@ -0,0 +1,367 @@
+   [toolbar/content-title (i18n/label :t/ens-your-username)]])
+
+(views/defview register []
+  (views/letsubs [{:keys [registrar]}          [:get-screen-params :ens-register]
```
make just one subscription for the register screen with all the data you need

yenda commented on src/status_im/ui/screens/ens/events.cljs:
```diff
@@ -0,0 +1,123 @@
+    (resolver/pubkey registry name
+                     (fn [ss]
+                       (if (= ss public-key)
+                         (re-frame/dispatch [:ens/set-state :connected])
```
don't do double dispatch, do one callback event that does these two things

jeluard replied:
Which 2 things?

yenda replied:
I missed the if

yenda commented on src/status_im/ui/screens/ens/events.cljs:
```diff
@@ -0,0 +1,123 @@
+
+(defn- chain [{:keys [network] :as db}]
+  (let [network (get-in db [:account/account :networks network])]
+    (ethereum/network->chain-keyword network)))
```
this is already done by ethereum/chain-keyword which takes db, so you don't actually need this functions

yenda commented on src/status_im/ui/screens/events.cljs:
```diff
@@ -9,6 +9,7 @@
             status-im.web3.events
             status-im.ui.screens.add-new.new-chat.navigation
             status-im.ui.screens.profile.events
+            status-im.ui.screens.ens.events
```
these are all wrong and should be refactored at some point, don't add a new one

antdanchenko committed:
added test_open_blocked_site

Signed-off-by: Anton Danchenko <ant.danchenko@gmail.com>

Commit:
[fix] remove duplicates of :hardwallet/verify-pin event

Signed-off-by: yenda <eric@status.im>

jakubgs committed:
add MOBILE_UI_FOR_DESKTOP=0 to other env files

Signed-off-by: Jakub Sokołowski <jakub@status.im>

dmitryn committed:
save keycard key-uid to account and use it to login

Signed-off-by: Dmitry Novotochinov <dmitry.novot@gmail.com>

Commit:
[dev feature] extend the defn macro

the defn macro knows register the function as an events under the
keywords provided in the :events key of the attributes map. It also
adds the interceptors provided in the :interceptors map

exemple:

```clojure
(fx/defn hello4
  {:doc "this function is useless as well"
   :events [:test/valid1 :test/valid2]}
  [{:keys [db]} b]
  {:db (assoc db :a b) :b (:a db)})
```

Signed-off-by: yenda <eric@status.im>

rasom committed:
[#8277] Report a bug template

Signed-off-by: Jakub Sokołowski <jakub@status.im>

flexsurfer commented:
![image](https://user-images.githubusercontent.com/11790366/58406452-ed242080-8069-11e9-9278-bffed6f8b4f5.png)

flexsurfer committed:
fix custom tokens

Signed-off-by: Andrey Shovkoplyas <motor4ik@gmail.com>

errorists commented:
Thanks Andrey for the Android demo. Some things I've noticed: 

👉 On Android, whenever we bold a font, it's no longer using the Inter font but rather the system default Roboto, not an issue contained to Julien's work, rather making a general note. cc @yenda because he loves fonts :P

![Group5](https://user-images.githubusercontent.com/4932774/58421600-a3026580-8090-11e9-9d4b-6cfcd4152553.png)

- [ ] layout here needs some adjustments, especially considering smaller screens. The empty spaces are left too large, content could use additional vertical centring

![Group](https://user-images.githubusercontent.com/4932774/58420309-39348c80-808d-11e9-86ab-98c681eb65c7.png)

- [ ] the text and button should be vertically centred

![Group2](https://user-images.githubusercontent.com/4932774/58420313-3cc81380-808d-11e9-8617-b36305d4690e.png)

- [ ] a small one, but both the top and bottom margin should have the same value here

![Group3](https://user-images.githubusercontent.com/4932774/58420314-3df94080-808d-11e9-948f-63026533818f.png)

- [ ] is there an underline here? if yes pls remove it 🙏

flexsurfer commented:
![image](https://user-images.githubusercontent.com/11790366/58423988-af89bc80-8096-11e9-8cb8-e70b567fb42c.png)

Commit:
Enable asset in Wallet before usage

Signed-off-by: Serhy <sergii@status.im>

Commit:
[fix] superfluous keyword

Signed-off-by: yenda <eric@status.im>

Commit:
[fix] on-result event for in app transactions

Signed-off-by: yenda <eric@status.im>

Commit:
Update Play Store phone screenshots

Signed-off-by: Igor Mandrigin <i@mandrigin.ru>

churik committed:
e2e pairing 1-1 and contacts

Signed-off-by: Churikova Tetiana <churikova.tm@gmail.com>

mandrigin committed:
Prepare to use status-go based protocol: 1-1 and public chat management APIs.

Connect to stubs of status-go protocol API, behind the flag. Since status-go isn't updated yet, setting this flag will break the app.
What needs to be tested is no regressions in a normal mode.

Signed-off-by: Igor Mandrigin <i@mandrigin.ru>

mandrigin committed:
bump version to 0.14.0

Signed-off-by: Igor Mandrigin <i@mandrigin.ru>

jeluard committed:
Connect and persist usernames

yenda commented:
could you fixup all these commit it can't be reviewed in that state

yenda reviewed (changes requested):
- there is still an ens events namespace, the best solution is to move to ens.core (outside of ui) and use the fx/defn attribute map like this: `{:events [:this-in-an-event]}` to declare the events.
- duplicated ens subscription
- internal transactions should use eth-transaction-call

jeluard commented:
@yenda Not done, please don't review.
